File: src/solutions/2024/06/ts/main.ts
import { argv } from 'node:process';
import type { SolutionOptions } from 'src/options.ts';
import { run } from 'src/run.ts';

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function findStart(grid: number[][]): { x: number; y: number } {
   for (let y = 0; y < grid.length; y++) {
      for (let x = 0; x < grid[y].length; x++) {
         if (grid[y][x] === 2) {
            return { x, y };
         }
      }
   }
   throw new Error('Start not found');
}

function part1(input: string, _isTest: boolean): string {
   const grid: number[][] = [];
   grid.push(
      ...input.split('\n').map((str) =>
         [3].concat(
            str.split('').map((x) => (x === '#' ? 1 : x === '^' ? 2 : 0)),
            [3],
         )
      ),
   );
   grid.unshift(new Array(grid[0].length).fill(3));
   grid.push(new Array(grid[0].length).fill(3));
   const visited = grid.map(() => new Array(grid[0].length).fill(false));

   const location = findStart(grid);
   const direction = { x: 0, y: -1 };

   while (true) {
      visited[location.y][location.x] = true;
      const xNext = location.x + direction.x;
      const yNext = location.y + direction.y;

      if (grid[yNext][xNext] === 3) {
         break;
      }

      if (grid[yNext][xNext] === 1) {
         if (direction.y === -1) {
            direction.x = 1;
            direction.y = 0;
         } else if (direction.x === 1) {
            direction.x = 0;
            direction.y = 1;
         } else if (direction.y === 1) {
            direction.x = -1;
            direction.y = 0;
         } else if (direction.x === -1) {
            direction.x = 0;
            direction.y = -1;
         }
      } else {
         location.x = xNext;
         location.y = yNext;
      }
   }
   return visited
      .flat()
      .filter((v) => v)
      .length.toString();
}

function part2(input: string, _isTest: boolean): string {
   const grid: number[][] = [];
   grid.push(
      ...input.split('\n').map((str) =>
         [3].concat(
            str.split('').map((x) => (x === '#' ? 1 : x === '^' ? 2 : 0)),
            [3],
         )
      ),
   );
   grid.unshift(new Array(grid[0].length).fill(3));
   grid.push(new Array(grid[0].length).fill(3));

   const visited = grid.map(() => new Array(grid[0].length).fill(false));
   const location = findStart(grid);
   const direction = { x: 0, y: -1 };

   const toVisit = [];
   while (true) {
      if (!visited[location.y][location.x]) {
         toVisit.push([location.x, location.y]);
      }
      visited[location.y][location.x] = true;
      const xNext = location.x + direction.x;
      const yNext = location.y + direction.y;

      if (grid[yNext][xNext] === 3) {
         break;
      }

      if (grid[yNext][xNext] === 1) {
         if (direction.y === -1) {
            direction.x = 1;
            direction.y = 0;
         } else if (direction.x === 1) {
            direction.x = 0;
            direction.y = 1;
         } else if (direction.y === 1) {
            direction.x = -1;
            direction.y = 0;
         } else if (direction.x === -1) {
            direction.x = 0;
            direction.y = -1;
         }
      } else {
         location.x = xNext;
         location.y = yNext;
      }
   }
   const maxVisit = toVisit.length * 2;

   const locationOg = findStart(grid);
   let count = 0;
   for (const v of toVisit) {
      const [x, y] = v;
      if (!visited[y][x] || grid[y][x] === 1 || grid[y][x] === 2) {
         continue;
      }

      grid[y][x] = 1;
      const location = { ...locationOg };
      const direction = { x: 0, y: -1 };
      let maxIteration = maxVisit;
      while (maxIteration) {
         const xNext = location.x + direction.x;
         const yNext = location.y + direction.y;

         if (grid[yNext][xNext] === 3) {
            break;
         }

         if (grid[yNext][xNext] === 1) {
            if (direction.y === -1) {
               direction.x = 1;
               direction.y = 0;
            } else if (direction.x === 1) {
               direction.x = 0;
               direction.y = 1;
            } else if (direction.y === 1) {
               direction.x = -1;
               direction.y = 0;
            } else if (direction.x === -1) {
               direction.x = 0;
               direction.y = -1;
            }
         } else {
            location.x = xNext;
            location.y = yNext;
         }
         maxIteration--;
      }
      if (maxIteration === 0) {
         count++;
      }
      grid[y][x] = 0;
   }
   return count.toString();
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}

File: src/solutions/2024/06/csharp/Main.cs

using Core;

namespace Year2024;
public static class Day06
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = false,
      HasIO = false,
   };

   static private Coordinate FindStart(char[][] grid)
   {
      for (int y = 0; y < grid.Length; y++)
      {
         for (int x = 0; x < grid[0].Length; x++)
         {
            if (grid[y][x] == '^')
            {
               return new Coordinate { x = x, y = y };
            }
         }
      }
      throw new Exception("Start not found");
   }

   struct Coordinate
   {
      public int x;
      public int y;
   }

   static string Part1(string input, bool isTest = false)
   {
      var gridL = input.Split('\n').Select(line =>
      {
         var l = line.ToList();
         l.Insert(0, 'x');
         l.Add('x');
         return l.ToArray();
      }).ToList();
      gridL.Insert(0, Enumerable.Repeat('x', gridL[0].Length).ToArray());
      gridL.Add(Enumerable.Repeat('x', gridL[0].Length).ToArray());
      var grid = gridL.Select(line => line.ToArray()).ToArray();
      var visited = new bool[grid.Length][];
      for (int i = 0; i < grid.Length; i++)
      {
         visited[i] = new bool[grid[0].Length];
      }

      var location = FindStart(grid);
      (int x, int y)[] direction = { (0, -1), (1, 0), (0, 1), (-1, 0) };
      var directionIndex = 0;

      while (true)
      {
         visited[location.y][location.x] = true;
         var xNext = location.x + direction[directionIndex].x;
         var yNext = location.y + direction[directionIndex].y;

         if (grid[yNext][xNext] == 'x') break;

         if (grid[yNext][xNext] == '#')
         {
            directionIndex++;
            directionIndex %= direction.Length;
         }
         else
         {
            location.x = xNext;
            location.y = yNext;
         }
      }

      return visited.SelectMany(v => v).Where(v => v).Count().ToString();
   }

   static string Part2(string input, bool isTest = false)
   {
      var gridL = input.Split('\n').Select(line =>
      {
         var l = line.ToList();
         l.Insert(0, 'x');
         l.Add('x');
         return l.ToArray();
      }).ToList();
      gridL.Insert(0, Enumerable.Repeat('x', gridL[0].Length).ToArray());
      gridL.Add(Enumerable.Repeat('x', gridL[0].Length).ToArray());
      var grid = gridL.Select(line => line.ToArray()).ToArray();
      var visited = new bool[grid.Length][];
      for (int i = 0; i < grid.Length; i++)
      {
         visited[i] = new bool[grid[0].Length];
      }

      (int x, int y)[] direction = [(0, -1), (1, 0), (0, 1), (-1, 0)];
      var directionIndex = 0;
      var location = FindStart(grid);
      while (true)
      {
         visited[location.y][location.x] = true;
         var xNext = location.x + direction[directionIndex].x;
         var yNext = location.y + direction[directionIndex].y;

         if (grid[yNext][xNext] == 'x') break;

         if (grid[yNext][xNext] == '#')
         {
            directionIndex++;
            directionIndex %= direction.Length;
         }
         else
         {
            location.x = xNext;
            location.y = yNext;
         }
      }

      var maxVisit = visited.SelectMany(v => v).Where(v => v).Count() * 2;
      var locationOg = FindStart(grid);
      var count = 0;
      for (int y = 1; y < grid.Length - 1; y++)
      {
         for (int x = 1; x < grid[0].Length - 1; x++)
         {
            if (!visited[y][x] || grid[y][x] == '#' || grid[y][x] == '^') continue;

            grid[y][x] = '#';
            location = new Coordinate { x = locationOg.x, y = locationOg.y };
            directionIndex = 0;
            var maxIteration = maxVisit;
            while (maxIteration > 0)
            {
               var xNext = location.x + direction[directionIndex].x;
               var yNext = location.y + direction[directionIndex].y;

               if (grid[yNext][xNext] == 'x') break;

               if (grid[yNext][xNext] == '#')
               {
                  directionIndex++;
                  directionIndex %= direction.Length;
               }
               else
               {
                  location.x = xNext;
                  location.y = yNext;
               }
               maxIteration--;
            }
            if (maxIteration == 0)
            {
               count++;
            }
            grid[y][x] = '.';
         }
      }
      return count.ToString();
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}
